engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/PartSubGraphCompiler.scala

```diff
@@ -51,24 +53,23 @@ class PartSubGraphCompiler(protected val classLoader: ClassLoader,
 
   protected def createServiceInvoker(obj: ObjectWithMethodDef) = ServiceInvoker(obj)
 
-  private val globalVariableTypes = expressionConfig.globalVariables.mapValues(_.returnType)
-
   def compile(n: SplittedNode[_], ctx: ValidationContext) : CompilationResult[compiledgraph.node.Node] = {
     implicit val nodeId: NodeId = NodeId(n.id)
 
-    def toCompilationResult[T](validated: ValidatedNel[ProcessCompilationError, T]) =
-      CompilationResult(Map(n.id -> ctx), validated)
+    def toCompilationResult[T](validated: ValidatedNel[ProcessCompilationError, T], expressionsTypingInfo: Map[String, ExpressionTypingInfo]) =
```

Comment:
would Writer help here or it's not worth it?

Reply:
I thought about using Writer here and in `Typer` but IMO disadvantages outweigh the advantages. To use it I should use some kind of MonadTransformer to compose effects of Validated and Writer. IMO It would be too complicated for scala dev with medium knowledge. I think that adding such an abstraction give the best benefits if this pattern is shared between many classes. In this case it is very local thing.

Reply:
probably you're right, although this method is already far too complicated for mid scala developer :P

Reply:
Yep, but I see no perfect solution for that. Also I'm not pretty sure if further polishing is worth doing it, because maybe in some near future we can do some general simplifications that will throw this code away like:
- Replacing all specified nodes like source, sink, filter and so on with simple "Node" with some `ContextTransformation`
- Removing `CanonicalProcess` layer
- Replacing PartSubGraphCompiler by some optional Optimizer that can be turned on or not.